rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,380 @@
+            token_address = out_event.asset.resolve_to_evm_token().evm_address
+        except WrongAssetType:
+            token_address = None
+
+        if token_address is None:
+            log.error(
+                f'Failed to decode swap from token to native token '
+                f'for tranction {transaction.tx_hash.hex()} due the or token_address is null.',
+            )
+            return
+
+        sell_amount = swap_amount = None
+        for tx_log in all_logs:
+            if (
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                tx_log.address == token_address
+            ):
+                # Find the transfer log from the user wallet to router
+                if bytes_to_address(tx_log.topics[2]) == self.router_address:
+                    sell_amount = int.from_bytes(tx_log.data)
+
+                # Find the transfer log from the router to the swap provider
+                elif bytes_to_address(tx_log.topics[1]) == self.router_address:
+                    swap_amount = int.from_bytes(tx_log.data)
+
+                if swap_amount and sell_amount:
+                    break
+
+        if not (sell_amount and swap_amount):
+            log.error(
+                f'Failed to decode swap from token to token '
+                f'for tranction {transaction.tx_hash.hex()} due the sell_amount '
+                'or swap_amount being null.',
+            )
+            return
+
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            timestamp=transaction.timestamp,
+            fee_raw=sell_amount - swap_amount,
+            fee_asset=out_event.asset.resolve_to_evm_token(),
+            sender=sender,
+        )
+        decoded_events.append(fee_event)
+
+        return
+
+    def _process_swap(self,
+            transaction: 'EvmTransaction',
+            decoded_events: list['EvmEvent'],
+            all_logs: list['EvmTxReceiptLog'],
+    ) -> list['EvmEvent']:
+        sender = transaction.from_address
+        in_event, out_event = self._get_in_out_events(decoded_events, sender)
+        if not (in_event and out_event):
+            log.error(
+                f'Failed to post decoding rainbow {transaction.tx_hash.hex()} '
+                f'due to missing in_event or out_event.',
+            )
+            return decoded_events
+
+        if out_event.asset == self.evm_inquirer.native_token and in_event.asset.is_evm_token():
+            # Try to find if have wrap eth into weth log
+            wrap_log = None
+            for tx_log in all_logs:
+                if tx_log.topics[0] == WETH_DEPOSIT_TOPIC:
+                    wrap_log = tx_log
+                    break
+            else:
+                log.error(
+                    f'Failed to decode swap from native token to token '
+                    f'for transaction {transaction.tx_hash.hex()} due to missing wrap log.',
+                )
+                return decoded_events
```

Comment:
this looks really similar to the other branch of the code. Can you try to abstract it? Maybe with a funtion that returns the event or None and you do something like

```
if (wrap_event := your_function(...)) is None:
     return decoded_events
```